MCLOUD-6135: Every step should throw only StepException

NadiyaS opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios,
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.

    If this PR was initiated by an internal JIRA ticket, copy the description from the ticket here so that the information is visible to contributors from the Magento Cloud Community.
-->
Every scenario step should throw only StepException.
If any error occurred on deploy phase - failed flag should be set and post-deploy skipped.

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/ece-tools#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. https://jira.corp.magento.com/browse/MCLOUD-6135

### Manual testing scenarios
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. ssh to cloud environment and set read only permissions to file app/etc/env.php `chmod 444 app/etc/env.php`
2. run redeployment
Expected result:
 - deploy phase is failed and flag .deploy_failed is set. Post-deploy script is skipped.

### Release notes

For user-facing changes, add a meaningful release note. For examples, see [Magento Cloud ECE-tools release notes](https://devdocs.magento.com/cloud/release-notes/ece-release-notes.html).

### Associated documentation updates
<!--
 If your proposed update requires user documentation, submit a PR to the Magento DevDocs repository. For extensive updates requiring assistance, submit an issue to DevDocs. See https://github.com/magento/devdocs/blob/master/.github/CONTRIBUTING.md.
 -->
Add link to Magento DevDocs PR or Issue, if needed.

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] Pull request introduces user-facing changes and includes meaningful updates for any required release notes and documentation changes
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

NadiyaS committed:
MCLOUD-6045: Replace shell_exec() function.

NadiyaS committed:
MCLOUD-6045: Replace shell_exec() function.

NadiyaS committed:
MCLOUD-6135: Every step should throw only StepException

NadiyaS committed:
MCLOUD-6135: Every step should throw only StepException

NadiyaS committed:
MCLOUD-6135: Every step should throw only StepException

NadiyaS committed:
MCLOUD-6135: Every step should throw only StepException

NadiyaS committed:
MCLOUD-6135: Every step should throw only StepException

NadiyaS committed:
MCLOUD-6135: Every step should throw only StepException

NadiyaS committed:
MCLOUD-6135: Every step should throw only StepException

NadiyaS committed:
MCLOUD-6135: Every step should throw only StepException

NadiyaS committed:
MCLOUD-6135: Every step should throw only StepException

NadiyaS committed:
MCLOUD-6135: Every step should throw only StepException

oshmyheliuk commented on src/Scenario/Processor.php:
```diff
@@ -64,19 +64,26 @@ public function execute(array $scenarios): void
             });
-        } catch (StepException $stepException) {
+        } catch (Throwable $stepException) {
             try {
```
Let's move nested try/catch block to a new separate method as this method contains too many try/catch blocks and it becomes difficult to read.
```
        } catch (Throwable $stepException) {
            $this->runActions();
            $this->handleException($stepException);
        }
```

NadiyaS replied:
done

oshmyheliuk commented on src/Scenario/Processor.php:
```diff
@@ -64,19 +64,26 @@ public function execute(array $scenarios): void
+                $exception,
+                sprintf('Unhandled error: [%d] %s', $exception->getCode(), $exception->getMessage())
+            );
+        }
```
Please add a newline to separate try/catch blocks

NadiyaS replied:
done

oshmyheliuk commented on src/Step/SkipStep.php:
```diff
@@ -39,9 +39,14 @@ public function __construct(LoggerInterface $logger, string $stepName)
     public function execute()
     {
-        $this->logger->info(sprintf('Step "%s" was skipped', $this->stepName));
+        try {
```
Why did you add try/catch here? Logger method `info` doesn't throw any exceptions.

NadiyaS replied:
changes reverted

oshmyheliuk commented on src/Step/SkipStep.php:
```diff
@@ -39,9 +39,14 @@ public function __construct(LoggerInterface $logger, string $stepName)
      * Logs the information about step skipping.
      *
      * @return void
+     * @throws StepException
```
these two lines can be changed to `{@inheritDoc}`

NadiyaS replied:
done

NadiyaS committed:
MCLOUD-6135: Every step should throw only StepException

NadiyaS committed:
MCLOUD-6135: Every step should throw only StepException

oshmyheliuk committed:
Merge branch 'develop' into MCLOUD-6135

BaDos committed:
Merge branch 'develop' into MCLOUD-6135

andriyShevtsov commented:
QA approved

NadiyaS committed:
Merge branch 'develop' of github.com:magento/ece-tools into MCLOUD-6135

# Conflicts:
#	src/Scenario/Processor.php
#	src/Step/Deploy/InstallUpdate/ConfigUpdate/DbConnection.php
#	src/Step/Deploy/SplitDbConnection.php

BarnyShergold commented on src/Step/Deploy/SplitDbConnection.php:
```diff
@@ -110,56 +110,60 @@ public function __construct(
-            $this->logger->warning(
+            if (!empty($notAvailableSplitTypes)) {
+                $this->logger->warning(
                 sprintf(
```
Indent of lines 133/137/138 causing tests to fail - should be indented a further 4 spaces.

NadiyaS replied:
fixed

andriyShevtsov replied:
@BarnyShergold please approve. Requested changes were implemented

NadiyaS committed:
MCLOUD-6135: Every step should throw only StepException

# Conflicts:
#	src/Scenario/Processor.php
#	src/Step/Deploy/InstallUpdate/ConfigUpdate/DbConnection.php
#	src/Step/Deploy/SplitDbConnection.php

NadiyaS committed:
Merge branch 'develop' of github.com:magento/ece-tools into MCLOUD-6135

# Conflicts:
#	src/Step/Deploy/InstallUpdate/ConfigUpdate/DbConnection.php
#	src/Step/Deploy/SplitDbConnection.php

oshmyheliuk commented on src/Scenario/Processor.php:
```diff
@@ -64,43 +63,55 @@ public function execute(array $scenarios): void
-            } catch (ActionException $actionException) {
-                $this->logger->error($actionException->getMessage());
-            }
+        } catch (Throwable $stepException) {
```
Looks like here should be `StepException` instead of `Throwable`, as `Step` can throw only `StepException`

oshmyheliuk replied:
you can change this method to something like this with one Try/Catch block
```php
    public function execute(array $scenarios): void
    {
        try {
            $this->logger->info(sprintf(
                'Starting scenario(s): %s %s',
                implode(', ', $scenarios),
                $this->packageManager->getPrettyInfo()
            ));

            $this->mergedScenarios = $this->merger->merge($scenarios);
            $steps = $this->mergedScenarios['steps'];

            array_walk($steps, function (StepInterface $step, string $name) {
                $this->logger->debug('Running step: ' . $name);

                $step->execute();

                $this->logger->debug(sprintf('Step "%s" finished', $name));
            });

            $this->logger->info('Scenario(s) finished');
        } catch (StepException $stepException) {
            try {
                $actions = $this->mergedScenarios['actions'];

                array_walk($actions, function (ActionInterface $action, string $name) {
                    $this->logger->debug('Running on fail action: ' . $name);

                    $action->execute();

                    $this->logger->debug(sprintf('On fail action "%s" finished', $name));
                });
            } catch (ActionException $actionException) {
                $this->logger->error($actionException->getMessage());
            }
            $this->handleException($stepException);
        } catch (Throwable $exception) {
            $this->handleException(
                $exception,
                sprintf('Unhandled error: [%d] %s', $exception->getCode(), $exception->getMessage())
            );
        }
    }
```

NadiyaS replied:
no, we should catch Throwable, otherwise there may be a case when we did not set flag that deploy was failed and post-deploy phase will be executed.

NadiyaS committed:
Merge branch 'develop' into MCLOUD-6135

andriyShevtsov commented:
QA approved

ghost commented:
Hi @NadiyaS, thank you for your contribution! </br>Please, complete [Contribution Survey](https://opensource.magento.com/survey/contribution/magento+ece-tools+745), it will take less than a minute. </br>Your feedback will help us to improve contribution process.
